src/main/java/build/buildfarm/instance/shard/ShardInstance.java

```diff
@@ -719,23 +727,33 @@ private ListenableFuture<Iterable<Digest>> findMissingBlobsQueryingEachWorker(
   // out-of-date and the server lies about which blobs are actually present. We provide this
   // alternative strategy for calculating missing blobs.
   private ListenableFuture<Iterable<Digest>> findMissingBlobsViaBackplane(
-      Iterable<Digest> nonEmptyDigests) {
+      Iterable<Digest> nonEmptyDigests, RequestMetadata requestMetadata) {
     try {
       Set<Digest> uniqueDigests = new HashSet<>();
       nonEmptyDigests.forEach(uniqueDigests::add);
       Map<Digest, Set<String>> foundBlobs = backplane.getBlobDigestsWorkers(uniqueDigests);
       Set<String> workerSet = backplane.getStorageWorkers();
       Map<String, Long> workersStartTime = backplane.getWorkersStartTimeInEpochSecs(workerSet);
-      return immediateFuture(
+      Map<Digest, Set<String>> digestAndWorkersMap =
           uniqueDigests.stream()
-              .filter( // best effort to present digests only missing on active workers
+              .map(
                   digest -> {
                     Set<String> initialWorkers =
                         foundBlobs.getOrDefault(digest, Collections.emptySet());
-                    return filterAndAdjustWorkersForDigest(
-                            digest, initialWorkers, workerSet, workersStartTime)
-                        .isEmpty();
+                    return Pair.of(
+                        digest,
+                        filterAndAdjustWorkersForDigest(
+                            digest, initialWorkers, workerSet, workersStartTime));
                   })
+              .collect(Collectors.toMap(Pair::getKey, Pair::getValue));
+
+      leaseExtensionService.execute(
+          () -> extendLeaseForDigests(digestAndWorkersMap, requestMetadata));
```

Comment:
The use of a separate executor service for this should not be necessary. Use a cascading request and direct executor to do this with a future chain. You should also create a single grpc context detached from the current to handle all of those requests.

Reply:
(not necessary because you can throttle # of reqs with non-blocking calls through a semaphore)

Reply:
With directExecutor(), I need to make sure that `extendLease` method is called after `responseObserver.onCompleted()`.

I can think of one approach pass a `SettableFuture` and add listener to extendLease. After `responseObserver.onCompleted()` invocation mark `SettableFuture` complete. Are you ok with this, or do you have anything else in mind

```
instance.findMissingBlobs(digestList, requestMetadata, onCompleteFuture /* new argument*/)

....
repsonseObserver.onCompleted()
onCompleteFuture.set(..)
```

In findMissingBlobs

```
ListenableFuture<Iterables>  findMissingBlobs(...) {
....
onCompleteFuture.addListener(() -> extendLease(..), directExecutor())
}
```

Reply:
Incorporated feedback, can you please check.

Reply:
@werkt can you please take a look.